File: 3/3100/WaterBottlesII.py
/*
https://leetcode.com/problems/water-bottles-ii/?envType=daily-question&envId=2025-10-02
*/

class Solution:
    def maxBottlesDrunk(self, numBottles: int, numExchange: int) -> int:
        currentFullBottles=numBottles
        currentEmptyBottles=0
        resultantBottlesDrink=0
        while(True):
            if currentFullBottles==0 and currentEmptyBottles<numExchange:
                break
            if currentEmptyBottles>=numExchange:
                currentFullBottles+=1
                currentEmptyBottles-=numExchange
                numExchange+=1
            else:
                currentEmptyBottles+=currentFullBottles
                resultantBottlesDrink+=currentFullBottles
                currentFullBottles=0
        return resultantBottlesDrink

File: 3/3000s/3100/WaterBottlesII.cpp
/*
https://leetcode.com/problems/water-bottles-ii/?envType=daily-question&envId=2025-10-02
*/

class Solution {
public:
    int maxBottlesDrunk(int numBottles, int numExchange) {
        int currFullBt=numBottles, currEmptyBt=0, res=0;
        while(true){
            if(currFullBt==0&&currEmptyBt<numExchange) break;
            if(currEmptyBt>=numExchange){
                currFullBt+=1;
                currEmptyBt-=numExchange;
                numExchange+=1;
            } else{
                currEmptyBt+=currFullBt;
                res+=currFullBt;
                currFullBt=0;
            }
        }
        return res;
    }
};